### PR TITLE
Seasonal model

### PR DESCRIPTION
**Pull request by keturner8**
In trend.py the function seasonal_model creates a least-squares fit of a timeseries using constant, linear, annual harmonic, and semi-annual harmonic terms. The function returns the residuals from the fit, and if requested, the seasonal model itself, as xarray DataArrays. The DataArrays are set up to have similar attributes to the original DataArrays.

In test_trend.py a check was added for this new function.

### CODE REVIEW

File: src/momlevel/trend.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 """ trend.py - utilities for working with trends """
 
 import warnings
-
+import numpy as np
 import xarray as xr
 
 __all__ = [
```
Comment (jkrasting) [inline src/momlevel/trend.py:7]: Please add `"seasonal_model",`to the `__all__` list.

File: src/momlevel/trend.py [hunk 2]
```diff
@@ -342,3 +340,108 @@ def linear_detrend(xobj, dim="time", order=1, mode="remove"):
...
+    pmodel_da = xr.DataArray(
+        pmodel,
+        dims=["time", "coeff"],
+        coords={"coeff": np.arange(1, 7, 1), "time": da_timeseries[tcoord]},
+    )
+    pmodel_da = pmodel_da.expand_dims(dim=hashable_coords)
+
+    mcoeff = pmodel_da.dot(da_timeseries, dims="time")
+
+    seasonal_model = model_da.dot(mcoeff, dims="coeff")
```
Comment (jkrasting) [inline src/momlevel/trend.py:415]: Using `seasonal_model` as a variable name when the function itself is named `seasonal_model` is not advised

File: src/momlevel/trend.py [hunk 3]
```diff
@@ -342,3 +340,108 @@ def linear_detrend(xobj, dim="time", order=1, mode="remove"):
...
+    # Here we find the non-time coordinates for our dataset and create a hashable
+    # so that the model can be expanded to an arbitrary number of dimensions
+    # If coordinates are empty we drop them
+    da_timeseries = da_timeseries.reset_coords(drop=True)
+    coords = [x for x in da_timeseries.coords if x != tcoord]
+    coords = tuple(coords)
+
+    coords_dict = {}
+    for i in range(len(coords)):
+        coords_dict[coords[i]] = da_timeseries[f"{coords[i]}"]
```
Comment (jkrasting) [inline src/momlevel/trend.py:378]: From pylint: Consider using enumerate instead of iterating with range and len

File: src/momlevel/trend.py [hunk 4]
```diff
@@ -342,3 +340,108 @@ def linear_detrend(xobj, dim="time", order=1, mode="remove"):
...
+        _units = ""
+
+    seasonal_model.attrs["standard_name"] = _standard_name_m
+    seasonal_model.attrs["long_name"] = _long_name_m
+    seasonal_model.attrs["units"] = _units
+
+    residuals.attrs["standard_name"] = _standard_name_r
+    residuals.attrs["long_name"] = _long_name_r
+    residuals.attrs["units"] = _units
+    if return_model == True:
```
Comment (jkrasting) [inline src/momlevel/trend.py:444]:
```suggestion
    if return_model:
```

**Review by jkrasting — CHANGES_REQUESTED** [review]
Looks really good, @keturner8!   Just a few minor nits.

Consider using `pylint` to check for these kinds of issues:

```
pip install pylint
pylint src/momlevel
```

File: src/momlevel/trend.py (hunk 1 above)
Comment (keturner8) [reply src/momlevel/trend.py:7]: Added in the latest commit

File: src/momlevel/trend.py (hunk 2 above)
Comment (keturner8) [reply src/momlevel/trend.py]: Changed in the latest commit

File: src/momlevel/trend.py (hunk 4 above)
Comment (keturner8) [reply src/momlevel/trend.py]: Done!

**Review by jkrasting — APPROVED** [review]

**Pull request merged**